src/librustc_typeck/check/_match.rs

```diff
@@ -401,11 +402,16 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                        match_src: hir::MatchSource) -> Ty<'tcx> {
         let tcx = self.tcx;
 
-        // Not entirely obvious: if matches may create ref bindings, we
-        // want to use the *precise* type of the discriminant, *not* some
-        // supertype, as the "discriminant type" (issue #23116).
+        // Not entirely obvious: if matches may create ref bindings, we want to
+        // use the *precise* type of the discriminant, *not* some supertype, as
+        // the "discriminant type" (issue #23116).
+        //
+        // FIXME(tschottdorf): don't call contains_explicit_ref_binding, which
+        // is problematic as the HIR is being scraped, but ref bindings may be
+        // implicit after #42640. We need to make sure that pat_adjustments
+        // (once introduced) is populated by the time we get here.
```

Comment:
No they're not. The problem is that if we have something like
```Rust
    let &mut Foo(x) = f()[0];
```

Then if the pattern matches by reference, we want to match `f()[0]` as a lexpr, so we can't allow it to be coerced. But if the pattern matches by value, `f()[0]` is still syntactically a lexpr, but we *do* want to allow coercions.